src/cc/mod.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019, Cloudflare, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+use std::time::Instant;
+
+use crate::cc;
+use crate::recovery::Sent;
+
+// Congestion Control constants
+pub const INITIAL_WINDOW_PACKETS: usize = 10;
+
+pub const INITIAL_WINDOW: usize = INITIAL_WINDOW_PACKETS * MAX_DATAGRAM_SIZE;
+
+pub const MINIMUM_WINDOW: usize = 2 * MAX_DATAGRAM_SIZE;
+
+pub const MAX_DATAGRAM_SIZE: usize = 1452;
+
+pub const LOSS_REDUCTION_FACTOR: f64 = 0.5;
+
+// Default congestion control algorithm
+pub const DEFAULT_CC_ALGORITHM: &str = cc::reno::CC_NAME_RENO;
+
+// Congestion Control Trait
+pub trait CongestionControlTrait {
+    fn new() -> Self
+    where
+        Self: Sized;
+
+    fn get_name(&self) -> &str;
+
+    fn to_string(&self) -> String {
+        self.get_name().to_string()
+    }
+
+    fn set_trace_id(&mut self, id: String);
+
+    // Access to internal variables
+    fn cwnd(&self) -> usize;
+
+    fn set_cwnd(&mut self, cwnd: usize);
+
+    fn ssthresh(&self) -> usize;
+
+    fn bytes_in_flight(&self) -> usize;
+
+    fn set_bytes_in_flight(&mut self, bytes_in_flight: usize);
```

Comment:
Actually, the only thing that actually reads and writes to this seems to be the recovery code, so maybe just leave the `bytes_in_flight` field in `Recovery` itself? So we can also get rid of these two methods.

Reply:
1. set_bytes_in_flight() is called in recovery.on_packets_lost() and recovery.drop_unacked_data(), so I exposed the API.
2. bytes_in_flight is originally mentioned in congestion control section in the draft, so I basically moved. Another issue is that if we have bytes_in_flight in Recovery it's hard to access from CC module, but Recovery can have access to CC module's variable in this way.

Reply:
Changed as follows. I think it's good enough for now.

recovery.set_bytes_in_flight() -> recovery.cc.set_bytes_in_flight()
recovery.bytes_in_flight() -> recovery.cc.bytes_in_flight()

Reply:
Still, I think we can expose a more specific API, rather than always having to do the `saturating_sub()` dance. Something like, a `drop_packet()` method, that does the subtraction internally seems better IMO.

Reply:
Ok. set_bytes_in_flight() is replaced with decrease_bytes_in_flight(usize). I think this is basically controlling bytes_in_flight variable, so better to have the name in its API name.